src/table.js

```diff
@@ -445,64 +450,168 @@ Table.prototype.createReadStream = function(options) {
+        return lessThan(rhs, lhs);
+      };
+      const greaterThanOrEqualTo = (lhs, rhs) => {
+        return !lessThan(rhs, lhs);
+      };
+
+      if (ranges.length === 0) {
+        ranges.push({
+          start: {
+            value: lastRowKey,
+            inclusive: false,
+          },
+        });
+      } else {
+        // Readjust and/or remove ranges based on previous valid row reads.
+
+        // Iterate backward since items may need to be removed.
+        for (let index = ranges.length - 1; index >= 0; index--) {
+          const range = ranges[index];
+          const normalizeRangeValue = rangeSide => {
+            return is.object(rangeSide) ? rangeSide.value : rangeSide;
+          };
+          const startValue = normalizeRangeValue(range.start);
+          const endValue = normalizeRangeValue(range.end);
+          const isWithinStart =
+            !startValue || greaterThanOrEqualTo(startValue, lastRowKey);
+          const isWithinEnd = !endValue || lessThan(lastRowKey, endValue);
+          if (isWithinStart) {
+            if (isWithinEnd) {
+              // The lastRowKey is within this range, adjust the start value.
+              range.start = {
+                value: lastRowKey,
+                inclusive: false,
+              };
+            } else {
+              // The lastRowKey is past this range, remove this range.
+              ranges.splice(index, 1);
+            }
+          }
+        }
       }
-      const row = self.row(rowData.key);
-      row.data = rowData.data;
-      next(null, row);
-    }),
-  ]);
-  return stream;
+
+      // Remove rowKeys already read.
+      if (rowKeys) {
+        rowKeys = rowKeys.filter(rowKey => greaterThan(rowKey, lastRowKey));
+        if (rowKeys.length === 0) {
+          rowKeys = null;
+        }
+      }
+    }
+    if (rowKeys || ranges.length) {
+      reqOpts.rows = {};
+      if (rowKeys) {
+        reqOpts.rows.rowKeys = rowKeys.map(Mutation.convertToBytes);
+      }
+      if (ranges.length) {
+        reqOpts.rows.rowRanges = ranges.map(function(range) {
+          return Filter.createRange(range.start, range.end, 'Key');
+        });
+      }
+    }
+    if (filter) {
+      reqOpts.filter = filter;
+    }
+
+    if (rowsLimit) {
+      reqOpts.rowsLimit = rowsLimit - rowsRead;
+    }
+
+    const requestStream = this.requestStream(grpcOpts, reqOpts);
+    requestStream.on('request', () => numRequestsMade++);
+
+    const stream = pumpify.obj([
+      requestStream,
+      chunkTransformer,
+      through.obj(function(rowData, enc, next) {
+        if (chunkTransformer._destroyed || retryStream._writableState.ended) {
+          return next();
+        }
+        numRequestsMade = 0;
+        rowsRead++;
+        const row = self.row(rowData.key);
+        row.data = rowData.data;
+        next(null, row);
+      }),
+    ]);
+
+    stream.on('error', error => {
+      // TODO - DO NOT MERGE - @stephenplusplus is this unpipe needed?
```

Comment:
Can you elaborate on the situation that would cause this, and what needs to happen?

Reply:
I'm not sure if an error on the `steam` will unpipe it from `retryStream`, considering that this stream could live for hours and multiple retried streams are piping to it, does this leak memory or add any significant overhead?

Reply:
I'm not sure what will happen internally. Calling unpipe should be safe to do.

Reply:
Cool, calling `.unpipe()` just to be sure.